ddev/src/ddev/cli/meta/scripts/upgrade_python.py

```diff
@@ -3,230 +3,391 @@
+WINDOWS_SHA_PATTERN = re.compile(
+    r'python-\$Env:PYTHON_VERSION-amd64\.exe[^\n]*\n[^\n]*-Hash\s+\'([0-9a-f]+)\'', re.MULTILINE
+)
+
+
+@click.command('upgrade-python-version', short_help='Upgrade the Python version used in the repository.')
 @click.pass_obj
-def upgrade_python(app: Application, version: str):
-    """Upgrade the Python version of all test environments.
+def upgrade_python_version(app: Application):
+    """Upgrade the Python version used in the repository.
+
+    Automatically detects the latest Python version, fetches official SHA256 hashes,
+    and updates version references across:
+    - ddev/src/ddev/repo/constants.py
+    - .builders/images/*/Dockerfile (Linux and Windows)
+    - .github/workflows/resolve-build-deps.yaml (macOS)
 
     \b
-    `$ ddev meta scripts upgrade-python 3.11`
+    `$ ddev meta scripts upgrade-python-version`
     """
+    from ddev.repo.constants import PYTHON_VERSION as major_minor
+    from ddev.repo.constants import PYTHON_VERSION_FULL as current_version
 
-    from ddev.repo.constants import PYTHON_VERSION as old_version
-
-    tracker = app.create_validation_tracker('Python upgrades')
+    tracker = app.create_validation_tracker('Python version upgrades')
 
-    for target in integrations(app):
-        update_hatch_file(app, target.path, version, old_version, tracker)
-        update_pyproject_file(target, version, old_version, tracker)
-        update_setup_file(target, version, old_version, tracker)
-
-    update_ci_files(app, version, old_version, tracker)
+    # Check for new version
+    latest_version = get_latest_python_version(app, major_minor)
+    if not latest_version:
+        app.display_error(f"Could not find latest Python version for {major_minor}")
+        app.abort()
+        return  # Unreachable but helps type checker
 
-    if app.repo.name == 'core':
-        update_ddev_pyproject_file(app, version, old_version, tracker)
-        update_constants_file(app, version, old_version, tracker)
-        update_ddev_template_files(app, version, old_version, tracker)
-        app.display_warning("Documentation files have not been updated. Please modify them manually.")
+    if Version(latest_version) <= Version(current_version):
+        app.display_info(f"Already at latest Python version: {current_version}")
+        return
 
-    tracker.display()
+    app.display_info(f"Updating Python from {current_version} to {latest_version}")
 
-    if tracker.errors:  # no cov
+    # Fetch and validate hashes (validation happens inside get_python_sha256_hashes)
+    try:
+        new_version_hashes = get_python_sha256_hashes(app, latest_version)
+    except Exception as e:
+        tracker.error(('SHA256 hashes',), message=f"Failed to fetch: {e}")
+        tracker.display()
         app.abort()
+        return  # Unreachable but helps type checker
 
+    # Perform updates
+    upgrade_python_version_full_constant(app, latest_version, tracker)
```

Comment:
nit. I guess would it make sense for the constant to be overridden after all the other files have been updated? Running in CI I think this is fine since it won't create a PR, but running locally and it fails, you'd have to manually reset the constants, I think.